File: src/stores/playlist.ts
import { create } from 'zustand';

export interface PlaylistItem {
  path: string;
  format: string;
  name: string;
}

export type PlayMode = 'single' | 'sequential' | 'loop' | 'loop_one' | 'random';

interface PlaylistStore {
  items: PlaylistItem[];
  currentIndex: number;
  playMode: PlayMode;

  setItems: (items: PlaylistItem[], currentIndex: number) => void;
  setCurrentIndex: (idx: number) => void;
  setPlayMode: (mode: PlayMode) => void;
  removeItem: (index: number) => void;
  moveItem: (from: number, to: number) => void;
}

/** Extract display name from path */
export function pathToName(path: string): string {
  const segs = path.split(/[/\\]/);
  const last = segs[segs.length - 1] ?? path;
  // strip extension
  const dot = last.lastIndexOf('.');
  return dot > 0 ? last.slice(0, dot) : last;
}

export const usePlaylistStore = create<PlaylistStore>((set) => ({
  items: [],
  currentIndex: -1,
  playMode: 'sequential',

  setItems: (items, currentIndex) => set({ items, currentIndex }),
  setCurrentIndex: (currentIndex) => set({ currentIndex }),
  setPlayMode: (playMode) => set({ playMode }),
  removeItem: (index) => set((s) => {
    const next = [...s.items];
    next.splice(index, 1);
    let ci = s.currentIndex;
    if (index < ci) ci--;
    else if (index === ci) ci = Math.min(ci, next.length - 1);
    return { items: next, currentIndex: next.length > 0 ? ci : -1 };
  }),
  moveItem: (from, to) => set((s) => {
    if (from < 0 || from >= s.items.length || to < 0 || to >= s.items.length || from === to) {
      return s;
    }
    const next = [...s.items];
    const [item] = next.splice(from, 1);
    next.splice(to, 0, item);
    // Track current index movement (mirror backend logic)
    let ci = s.currentIndex;
    if (ci === from) ci = to;
    else if (from < ci && to >= ci) ci--;
    else if (from > ci && to <= ci) ci++;
    return { items: next, currentIndex: ci };
  }),
}));

File: src-tauri/src/commands/playlist.rs
use tauri::State;
use std::path::PathBuf;

use crate::state::AppState;

/// Add files to playlist
#[tauri::command]
pub async fn playlist_add_files(
    state: State<'_, AppState>,
    paths: Vec<String>,
) -> Result<usize, String> {
    let mut playlist = state.playlist.lock();
    let pbs: Vec<PathBuf> = paths.iter().map(PathBuf::from).collect();
    playlist.active_mut().add_files(&pbs);
    let len = playlist.active().len();
    state.playlist_saver.mark_dirty();
    Ok(len)
}

/// Get playlist items
#[tauri::command]
pub async fn playlist_get_items(
    state: State<'_, AppState>,
) -> Result<serde_json::Value, String> {
    let playlist = state.playlist.lock();
    let items: Vec<_> = playlist.active().items
        .iter()
        .map(|item| serde_json::json!({
            "path": item.path,
            "format": format!("{:?}", item.format),
        }))
        .collect();
    Ok(serde_json::json!({ "items": items, "currentIndex": playlist.active().current_index }))
}

/// Play next track
#[tauri::command]
pub async fn playlist_next(
    state: State<'_, AppState>,
) -> Result<Option<String>, String> {
    let next_path = {
        let mut playlist = state.playlist.lock();
        let p = playlist.active_mut().next().map(|s| s.to_string());
        state.playlist_saver.mark_dirty();
        p
    };

    if let Some(ref path) = next_path {
        let mut player = state.player.lock();
        player.open_and_play(PathBuf::from(path).as_path())
            .map_err(|e| e.to_string())?;
    }

    Ok(next_path)
}

/// Play previous track
#[tauri::command]
pub async fn playlist_prev(
    state: State<'_, AppState>,
) -> Result<Option<String>, String> {
    let prev_path = {
        let mut playlist = state.playlist.lock();
        let p = playlist.active_mut().prev().map(|s| s.to_string());
        state.playlist_saver.mark_dirty();
        p
    };

    if let Some(ref path) = prev_path {
        let mut player = state.player.lock();
        player.open_and_play(PathBuf::from(path).as_path())
            .map_err(|e| e.to_string())?;
    }

    Ok(prev_path)
}

/// Play track at specific index
#[tauri::command]
pub async fn playlist_play_index(
    state: State<'_, AppState>,
    index: usize,
) -> Result<(), String> {
    let path = {
        let mut playlist = state.playlist.lock();
        playlist.active_mut().current_index = index;
        let p = playlist.active()
            .items
            .get(index)
            .map(|t| t.path.clone());
        state.playlist_saver.mark_dirty();
        p
    };

    if let Some(path) = path {
        let mut player = state.player.lock();
        player.open_and_play(PathBuf::from(path).as_path())
            .map_err(|e| e.to_string())?;
    }

    Ok(())
}

/// Clear playlist
#[tauri::command]
pub async fn playlist_clear(
    state: State<'_, AppState>,
) -> Result<(), String> {
    let mut playlist = state.playlist.lock();
    playlist.active_mut().clear();
    state.playlist_saver.mark_dirty();
    Ok(())
}

/// Remove track
#[tauri::command]
pub async fn playlist_remove(
    state: State<'_, AppState>,
    index: usize,
) -> Result<(), String> {
    let mut playlist = state.playlist.lock();
    playlist.active_mut().remove(index);
    state.playlist_saver.mark_dirty();
    Ok(())
}

/// Move a track within the playlist (reorder).
/// `from` and `to` are zero-based indices.
#[tauri::command]
pub async fn playlist_move_item(
    state: State<'_, AppState>,
    from: usize,
    to: usize,
) -> Result<(), String> {
    let mut playlist = state.playlist.lock();
    playlist.active_mut().move_item(from, to);
    state.playlist_saver.mark_dirty();
    Ok(())
}

/// Recursively scan a folder for audio files and add them to the playlist.
/// Returns the number of files added.
#[tauri::command]
pub async fn playlist_add_folder(
    state: State<'_, AppState>,
    folder: String,
) -> Result<usize, String> {
    let dir = PathBuf::from(&folder);
    if !dir.is_dir() {
        return Err(format!("Not a directory: {}", folder));
    }
    let mut playlist = state.playlist.lock();
    let count = playlist.active_mut().add_folder(&dir);
    state.playlist_saver.mark_dirty();
    Ok(count)
}

/// Get the current play mode of the active playlist.
#[tauri::command]
pub async fn playlist_get_play_mode(
    state: State<'_, AppState>,
) -> Result<String, String> {
    let playlist = state.playlist.lock();
    Ok(format!("{:?}", playlist.active().play_mode).to_lowercase())
}

/// Set the play mode of the active playlist.
#[tauri::command]
pub async fn playlist_set_play_mode(
    state: State<'_, AppState>,
    mode: String,
) -> Result<(), String> {
    let mode = match mode.to_lowercase().as_str() {
        "single" => tt_common::PlayMode::Single,
        "sequential" => tt_common::PlayMode::Sequential,
        "loop" => tt_common::PlayMode::Loop,
        "loop_one" | "loopone" => tt_common::PlayMode::LoopOne,
        "random" => tt_common::PlayMode::Random,
        other => return Err(format!("Unknown play mode: {}", other)),
    };
    let mut playlist = state.playlist.lock();
    playlist.active_mut().set_play_mode(mode);
    state.playlist_saver.mark_dirty();
    Ok(())
}
